### PR TITLE
Improve std::process module docs

### PR DESCRIPTION
**Pull request by Technius**
Addresses part of #29370

I've changed the first `cat` example to a "Hello World" example involving echo, and I've also added another example showing how to pipe output. I'm still working on the module-level description.

For now, I'd like feedback on the examples.

r? @steveklabnik

### CODE REVIEW

File: src/libstd/process.rs [hunk 1]
```diff
@@ -10,24 +10,57 @@
...
+//! Hello world, `std::process` edition:
 //!
-//! ```should_panic
+//! ```
 //! use std::process::Command;
 //!
-//! let mut child = Command::new("/bin/cat")
-//!                         .arg("file.txt")
+//! // Note that by default, the output of the command will be sent to stdout
+//! let child = Command::new("echo")
```
Comment (kennytm) [inline src/libstd/process.rs]: `child` needs to be `mut`.

```
[01:19:37] failures:
[01:19:37] 
[01:19:37] ---- process.rs - process (line 21) stdout ----
[01:19:37] 	error[E0596]: cannot borrow immutable local variable `child` as mutable
[01:19:37]   --> process.rs:12:13
[01:19:37]    |
[01:19:37] 7  | let child = Command::new("echo")
[01:19:37]    |     ----- consider changing this to `mut child`
[01:19:37] ...
[01:19:37] 12 | let ecode = child.wait()
[01:19:37]    |             ^^^^^ cannot borrow mutably
[01:19:37] 
[01:19:37] thread 'rustc' panicked at 'couldn't compile the test', /checkout/src/librustdoc/test.rs:283:12
[01:19:37] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:19:37] 
[01:19:37] 
[01:19:37] failures:
[01:19:37]     process.rs - process (line 21)
[01:19:37] 
[01:19:37] test result: FAILED. 877 passed; 1 failed; 10 ignored; 0 measured; 0 filtered out
```

File: src/libstd/process.rs [hunk 2]
```diff
@@ -10,22 +10,62 @@
...
+//! Several methods on [`Command`], such as [`spawn`] or [`output`], can be used
+//! to spawn a process. In particular, [`output`] spawns the child process and
+//! waits until the process terminates, while [`spawn`] will return a [`Child`]
+//! that represents the spawned child process.
+//!
+//! # Handling I/O
+//!
+//! TODO
+//!
+//! # Examples
```
Comment (steveklabnik) [inline src/libstd/process.rs]: why is this moved down? The above is still an example 😄

File: src/libstd/process.rs [hunk 3]
```diff
@@ -10,22 +10,62 @@
...
+//! ```
+//!
+//! Several methods on [`Command`], such as [`spawn`] or [`output`], can be used
+//! to spawn a process. In particular, [`output`] spawns the child process and
+//! waits until the process terminates, while [`spawn`] will return a [`Child`]
+//! that represents the spawned child process.
+//!
+//! # Handling I/O
+//!
+//! TODO
```
Comment (steveklabnik) [inline src/libstd/process.rs]: do you intend on filling this out in this PR? If not, I'd remove this section.

**Review by steveklabnik — COMMENTED** [review]
This looks good, but a few questions. Also travis is failing for some reason.

File: src/libstd/process.rs (hunk 3 above)
Comment (Technius) [reply src/libstd/process.rs]: Yes, I plan on adding a few demonstrations of `Stdio` here. Incidentally, the `TODO` is causing travis to fail.

File: src/libstd/process.rs (hunk 2 above)
Comment (Technius) [reply src/libstd/process.rs]: Good point! I planned on having a brief overview of the entire module first and then show some additional examples. Maybe I should rename the "Example" section to "Additional Examples"?

Comment (Technius) [reply src/libstd/process.rs]: I just figured out that I could just fold the examples into the section describing `Stdio`. I'll be updating the PR soon.

**Comment by Technius** [comment]
I think I've covered almost everything. If there's something else that I should mention in the docs, let me know!

**Comment by carols10cents** [comment]
Looks like this is back in @steveklabnik's court-- I think he's at conferences this week though :)

**Review by steveklabnik — APPROVED** [review]

**Comment by steveklabnik** [comment]
Yup, sorry that it took me some time. Looks great, thanks a ton!

@bors: r+ rollup

**Comment by kennytm** [comment]
@bors r- 

The example failed to run on Windows.

```
[01:31:48] ---- process.rs - process (line 76) stdout ----
[01:31:48] 	thread 'rustc' panicked at 'test executable failed:
[01:31:48] 
[01:31:48] thread 'main' panicked at 'failed to execute child: Error { repr: Os { code: 2, message: "The system cannot find the file specified." } }', src\libcore\result.rs:906:4
[01:31:48] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:31:48] 
[01:31:48] ', src\librustdoc\test.rs:323:16
[01:31:48] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:31:48] 
[01:31:48] 
[01:31:48] failures:
[01:31:48]     process.rs - process (line 76)
```

File: src/libstd/process.rs [hunk 4]
```diff
@@ -10,27 +10,68 @@
...
+//!     .expect("Failed to start sed process");
+//!
+//! let output = sed_child.wait_with_output().expect("Failed to wait on sed");
+//! assert_eq!(b"Oh no, a typo!\n", output.stdout.as_slice());
+//! ```
+//!
+//! Note that [`ChildStderr`] and [`ChildStdout`] implement [`Write`] and
+//! [`ChildStdin`] implements [`Read`]:
+//!
+//! ```
```
Comment (kennytm) [inline src/libstd/process.rs]: Windows does not have `/bin/cat`. Could you keep it ` ```no_run ` or change to another program for this example?

Comment (Technius) [reply src/libstd/process.rs]: I added `no_run` to the example involving `cat` and the example involving `sed`.
